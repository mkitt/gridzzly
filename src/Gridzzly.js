// @flow
import React, { PureComponent } from 'react'

const isListenable = props => (
  (typeof document !== 'undefined') && (props.cycleKey != null || props.toggleKey != null)
)

const defaultSize = (wanted, size) => (
  wanted != null ? wanted : size
)

const getSizes = props => ({
  cs: defaultSize(props.columnSize, props.size),
  rs: defaultSize(props.rowSize, props.size),
})

const getColumnLineOuter = (props) => {
  const { cs, rs } = getSizes(props)
  const dx = cs - (props.strokeWidth * 0.5)
  return `<line stroke='${props.colorOuter}' stroke-dasharray='${props.dashOuter}' x1='${dx}' y1='0' x2='${dx}' y2='${rs}'/>`
}

const getColumnLineInner = (props) => {
  const { cs, rs } = getSizes(props)
  const dx = (cs * 0.5) - (props.strokeWidth * 0.5)
  return `<line stroke='${props.colorInner}' stroke-dasharray='${props.dashInner}' x1='${dx}' y1='0' x2='${dx}' y2='${rs}'/>`
}

const getRowLineOuter = (props) => {
  const { cs, rs } = getSizes(props)
  const dy = rs - (props.strokeWidth * 0.5)
  return `<line stroke='${props.colorOuter}' stroke-dasharray='${props.dashOuter}' x1='0' y1='${dy}' x2='${cs}' y2='${dy}'/>`
}

const getRowLineInner = (props) => {
  const { cs, rs } = getSizes(props)
  const dy = (rs * 0.5) - (props.strokeWidth * 0.5)
  return `<line stroke='${props.colorInner}' stroke-dasharray='${props.dashInner}' x1='0' y1='${dy}' x2='${cs}' y2='${dy}'/>`
}

const getBase64GridString = ({ cs, rs, lines }) => (
  `data:image/svg+xml;charset=utf8,<svg xmlns='http://www.w3.org/2000/svg' width='${cs}' height='${rs}' viewBox='0 0 ${cs} ${rs}'><g fill='none' stroke-miter-limit='10'>${lines}</g></svg>`
)

const getLines = props => ([
  ...(props.showColumns && props.hasInner ? [getColumnLineInner(props)] : []),
  ...(props.showColumns ? [getColumnLineOuter(props)] : []),
  ...(props.showRows && props.hasInner ? [getRowLineInner(props)] : []),
  ...(props.showRows ? [getRowLineOuter(props)] : []),
].join())

const getStyle = (props) => {
  const lines = getLines(props)
  if (props.isDisabled || !lines.length) { return { display: 'none' } }
  const { cs, rs } = getSizes(props)
  const bgi = getBase64GridString({ cs, rs, lines })
  return {
    position: props.position,
    top: props.top,
    right: props.right,
    bottom: props.bottom,
    left: props.left,
    zIndex: props.zIndex,
    opacity: props.opacity,
    backgroundImage: `url("${bgi}")`,
    backgroundSize: `${cs}px ${rs}px`,
    pointerEvents: 'none',
    ...(props.style || {}),
  }
}

type Props = {
  autoHide?: boolean,
  columnSize?: ?number,
  colorInner?: string,
  colorOuter?: string,
  cycleKey?: ?string,
  dashInner?: string,
  dashOuter?: string,
  hasInner?: boolean,
  isDisabled?: boolean,
  opacity?: number,
  rowSize?: ?number,
  size?: number,
  strokeWidth?: number,
  toggleKey?: ?string,
  position?: 'absolute' | 'fixed' | 'sticky' | 'static',
  top?: number | string,
  right?: number | string,
  bottom?: number | string,
  left?: number | string,
  zIndex?: number,
}

type State = {
  showColumns: boolean,
  showRows: boolean,
}

export default class extends PureComponent<Props, State> {
  static defaultProps = {
    autoHide: false,
    colorInner: 'rgba(255, 0, 255, 0.75)',
    colorOuter: 'rgba(255, 0, 255, 1.0)',
    cycleKey: null,
    dashInner: '2, 2',
    dashOuter: '',
    hasInner: true,
    isDisabled: false,
    opacity: 0.2,
    size: 16,
    columnSize: null,
    rowSize: null,
    strokeWidth: 1,
    toggleKey: null,
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    zIndex: 2,
  }

  state = {
    showColumns: !(this.props.autoHide),
    showRows: !(this.props.autoHide),
  }

  componentDidMount() {
    if (isListenable(this.props)) {
      document.addEventListener('keydown', this.onKeyDown)
    }
  }

  componentWillUnmount() {
    if (isListenable(this.props)) {
      document.removeEventListener('keydown', this.onKeyDown)
    }
  }

  onKeyDown = (e: KeyboardEvent) => {
    const { cycleKey, toggleKey } = this.props
    if (toggleKey === e.key) {
      this.toggle()
    } else if (cycleKey === e.key) {
      this.cycle()
    }
  }

  cycle = () => {
    const { showRows, showColumns } = this.state
    if (showRows && showColumns) {
      this.setState({ showRows: false, showColumns })
    } else if (!showRows && !showColumns) {
      this.setState({ showRows, showColumns: true })
    } else if (showRows && !showColumns) {
      this.setState({ showRows: false, showColumns: true })
    } else if (!showRows && showColumns) {
      this.setState({ showRows: true, showColumns: false })
    }
  }

  toggle = () => {
    const { showRows, showColumns } = this.state
    if (showRows && showColumns) {
      this.setState({ showRows: false, showColumns: false })
    } else if (!showRows && !showColumns) {
      this.setState({ showRows: true, showColumns: true })
    } else if (showRows && !showColumns) {
      this.setState({ showRows, showColumns: true })
    } else if (!showRows && showColumns) {
      this.setState({ showRows: true, showColumns })
    }
  }

  render() {
    const elementProps = Object.assign({}, this.props)
    Object.keys(this.constructor.defaultProps).forEach((key) => {
      delete elementProps[key]
    })
    return (
      <div
        style={getStyle({ ...this.props, ...this.state })}
        {...elementProps}
      />
    )
  }
}
