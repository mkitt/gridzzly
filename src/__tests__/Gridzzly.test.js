import React from 'react'
import renderer from 'react-test-renderer'
import Gridzzly from '../Gridzzly'

describe('<Gridzzly />', () => {
  it('shows the correct snapshot tree for the <Gridzzly> component (fixed)', () => {
    const component = (
      <Gridzzly
        columnSize={32}
        cycleKey="~"
        position="fixed"
        rowSize={64}
        toggleKey="`"
        data-stuff="stuff"
      />
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('shows the correct snapshot tree for the <Gridzzly> component (defaults)', () => {
    const component = (
      <Gridzzly
        size={16}
        data-stuff="other-stuff"
      />
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('shows the correct snapshot tree for the <Gridzzly> component (hidden)', () => {
    const component = (
      <Gridzzly
        autoHide
      />
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
