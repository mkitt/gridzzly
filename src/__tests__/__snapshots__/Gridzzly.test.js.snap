// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gridzzly /> shows the correct snapshot tree for the <Gridzzly> component (defaults) 1`] = `
<div
  data-stuff="other-stuff"
  style={
    Object {
      "backgroundImage": "url(\\"data:image/svg+xml;charset=utf8,<svg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 16 16'><g fill='none' stroke-miter-limit='10'><line stroke='rgba(255, 0, 255, 0.75)' stroke-dasharray='2, 2' x1='7.5' y1='0' x2='7.5' y2='16'/>,<line stroke='rgba(255, 0, 255, 1.0)' stroke-dasharray='' x1='15.5' y1='0' x2='15.5' y2='16'/>,<line stroke='rgba(255, 0, 255, 0.75)' stroke-dasharray='2, 2' x1='0' y1='7.5' x2='16' y2='7.5'/>,<line stroke='rgba(255, 0, 255, 1.0)' stroke-dasharray='' x1='0' y1='15.5' x2='16' y2='15.5'/></g></svg>\\")",
      "backgroundSize": "16px 16px",
      "bottom": 0,
      "left": 0,
      "opacity": 0.2,
      "pointerEvents": "none",
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 2,
    }
  }
/>
`;

exports[`<Gridzzly /> shows the correct snapshot tree for the <Gridzzly> component (fixed) 1`] = `
<div
  data-stuff="stuff"
  style={
    Object {
      "backgroundImage": "url(\\"data:image/svg+xml;charset=utf8,<svg xmlns='http://www.w3.org/2000/svg' width='32' height='64' viewBox='0 0 32 64'><g fill='none' stroke-miter-limit='10'><line stroke='rgba(255, 0, 255, 0.75)' stroke-dasharray='2, 2' x1='15.5' y1='0' x2='15.5' y2='64'/>,<line stroke='rgba(255, 0, 255, 1.0)' stroke-dasharray='' x1='31.5' y1='0' x2='31.5' y2='64'/>,<line stroke='rgba(255, 0, 255, 0.75)' stroke-dasharray='2, 2' x1='0' y1='31.5' x2='32' y2='31.5'/>,<line stroke='rgba(255, 0, 255, 1.0)' stroke-dasharray='' x1='0' y1='63.5' x2='32' y2='63.5'/></g></svg>\\")",
      "backgroundSize": "32px 64px",
      "bottom": 0,
      "left": 0,
      "opacity": 0.2,
      "pointerEvents": "none",
      "position": "fixed",
      "right": 0,
      "top": 0,
      "zIndex": 2,
    }
  }
/>
`;

exports[`<Gridzzly /> shows the correct snapshot tree for the <Gridzzly> component (hidden) 1`] = `
<div
  style={
    Object {
      "display": "none",
    }
  }
/>
`;
